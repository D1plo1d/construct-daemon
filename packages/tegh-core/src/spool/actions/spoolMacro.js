import spoolTask from './spoolTask'
import { NORMAL } from '../types/PriorityEnum'
import runMacro from '../../config/selectors/runMacro'
import getMacroRunFn from '../../pluginManager/selectors/getMacroRunFn'

/*
 * spools the macro with the given args
 *
 * internal?: Boolean [default: false]
 * priority?: TaskPriority [default: macro.priority || NORMAL],
 * macro: String
 * args?: JSON [default: null]
 */
const spoolMacro = ({
  internal = false,
  priority,
  macro,
  args,
}) => (
  dispatch,
  getState,
) => {
  if (macro == null) {
    throw new Error('macro must not be null')
  }

  const state = getState()

  const gcodeLines = runMacro(state)(macro, args)

  const action = spoolTask({
    name: macro,
    internal,
    priority: priority || getMacroRunFn(state)(macro).priority || NORMAL,
    data: gcodeLines,
  })

  return dispatch(action)
}

export default spoolMacro
