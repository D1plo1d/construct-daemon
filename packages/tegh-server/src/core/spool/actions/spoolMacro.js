import spoolTask from './spoolTask'

/*
 * spools the macro with the given args
 *
 * internal?: Boolean [default: false]
 * priority?: TaskPriority [default: macro.priority || 'normal'],
 * macro: String
 * args?: JSON [default: null]
 */
const spoolMacro = ({ internal = false, priority, macro, args }) => {
  return (dispatch, getState) => {
    if (macro == null) {
      throw new Error('macro must not be null')
    }

    const state = getState()
    const macroDefinition = state.macros[macro]

    if (macroDefinition == null) {
      throw new Error(`Macro ${macro} does not exist`)
    }

    const gcode = macroDefinition.run(args, state.config)

    // TODO: move to a reducer
    // if (state.driver.status !== 'ready' && priority !== 'emergency') {
    //   throw new Error('Machine is not ready')
    // }

    const action = spoolTask({
      name: macroDefinition.name,
      internal,
      priority: priority || macroDefinition.priority || 'normal',
      data: gcode,
    })

    return dispatch(action)
  }
}

export default spoolMacro
