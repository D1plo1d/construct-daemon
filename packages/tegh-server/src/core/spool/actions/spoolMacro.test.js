import expectToMatchImmutableSnapshot from '../../util/testing/expectToMatchImmutableSnapshot'
import spoolMacro from './spoolMacro'

describe('spoolMacro', () => {
  it('creates a SPOOL_TASK action', async function() {
    const macro = 'my_test_macro'
    const args = {x: 360}

    const dispatch = action => action
    const getState = () => ({
      macros: {
        [macro]: {
          name: macro,
          priority: 'emergency',
          run: ({x}) => [`g1 x${x}`]
        }
      }
    })

    const result = await spoolMacro({
      macro,
      args,
    })(dispatch, getState)

    expectToMatchImmutableSnapshot({
      result,
      redactions: [
        ['payload', 'task', 'id'],
        ['payload', 'task', 'createdAt']
      ],
    })
  })
})
