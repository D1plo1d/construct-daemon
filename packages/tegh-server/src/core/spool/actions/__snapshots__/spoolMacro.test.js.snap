// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spoolMacro creates a SPOOL_TASK action 1`] = `
Object {
  "payload": Object {
    "task": Object {
      "createdAt": "[REDACTED]",
      "currentLineNumber": null,
      "data": Array [
        "g1 x360",
      ],
      "id": "[REDACTED]",
      "internal": false,
      "jobFileID": null,
      "jobID": null,
      "name": "my_test_macro",
      "priority": "EMERGENCY",
      "startedAt": null,
      "status": "SPOOLED",
      "stoppedAt": null,
    },
  },
  "type": "tegh-server/spool/SPOOL_TASK",
}
`;
